src/poll/responser.rs

```diff
@@ -0,0 +1,666 @@
+// TODO: this `#![allow(unused)]` lint is temporarily added. Will be removed sooner.
+#![allow(unused)]
+use std::collections::HashMap;
+use std::sync::RwLock;
+use std::{net::Ipv4Addr, slice::from_raw_parts_mut, sync::Arc, thread::spawn};
+
+use bitfield::bitfield;
+use libc::IFA_ADDRESS;
+use lockfree::queue::Queue;
+
+use crate::qp::QpContext;
+use crate::{device::QpType::RawPacket, Sge};
+use crate::{
+    device::{
+        ToCardWorkRbDescBuilder, ToCardWorkRbDescCommon, ToHostWorkRbDescAethCode,
+        ToHostWorkRbDescOpcode, ToHostWorkRbDescRead,
+    },
+    Device, Error,
+};
+
+/// A single MR that contains a free list of packet buffer. Each slot is 64 Byte.
+pub(crate) const ACKNOWLEDGE_BUFFER_SLOT: usize = 64;
+/// The size of the ack buffer.
+pub(crate) const ACKNOWLEDGE_BUFFER_SIZE: usize = ACKNOWLEDGE_BUFFER_SLOT * 1024;
+
+/// A interface that allows DescResponser to push the work descriptor to the device
+pub trait WorkDescriptorSender: Send + Sync {
+    fn push(&self, desc_builder: ToCardWorkRbDescBuilder) -> Result<(), Error>;
+}
+
+impl WorkDescriptorSender for Device {
```

Comment:
Did we implement too many functions for "Device"? What does a "push" action on Device mean?

And, now that Device already has a "send_work_desc", why need another trait simply do the same thing?

Reply:
To be honest, the first reason I created the interface for the `Device` is because it makes it easier to test the single module.  Another reason is that letting the `send_work_desc` as `pub` gives other structs the ability to visit that interface.  So I wonder should we separate it as an independence trait?

Reply:
I mean can you define `send_work_desc` in a trait so that we need only one trait?

Reply:
Got it. I will move the method `send_work_desc` to the `WorkDescriptorSender`.